File: 백준/Gold/14502. 연구소/연구소.cc
#include <bits/stdc++.h>

using namespace std;

#define X first
#define Y second

int board1[10][10];


int n, m;
int mx = -0x7f7f7f7f;

int wall[100];
int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };


int main(void) {
	ios::sync_with_stdio(0);
	cin.tie(0);

	cin >> n >> m;
	
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> board1[i][j];
		}
	}

	int idx = n * m;

	fill(wall + (idx - 3), wall + idx, 1);

	do {

		int board2[10][10];

		queue<pair<int, int>> Q = {};
		bool vis[10][10] = {};
		int cnt = 0;
		int ans = 0;

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				board2[i][j] = board1[i][j];
				if (board1[i][j] == 2) {
					Q.push({ i, j });
					vis[i][j] = 1;
				}
			}
		}

		for (int i = 0; i < n * m; i++) {
			int x = i / m;
			int y = i % m;

			if (wall[i] == 1) {
				if (board1[x][y] == 0){
					board2[x][y] = 1;
					cnt++;
				}
			}

		}

		if (cnt != 3) {
			continue;
		}

		while (!Q.empty()) {
			auto cur = Q.front(); Q.pop();
			for (int dir = 0; dir < 4; dir++) {
				int nx = cur.X + dx[dir];
				int ny = cur.Y + dy[dir];

				if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
				if (vis[nx][ny] || board2[nx][ny] != 0) continue;

				vis[nx][ny] = 1;
				Q.push({ nx, ny });
				board2[nx][ny] = 2;
			}
		}

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (board2[i][j] == 0)
				ans++;
			}
		}
		mx = max(mx, ans);

	} while (next_permutation(wall, wall + idx));

	cout << mx;
}

File: 백준/Gold/14502. 연구소/연구소.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Queue;
import java.util.StringTokenizer;
public class Main {

	static int N, M;
	static int[][] board;
	static int[][] cpy;
	static List<int[]> blank = new ArrayList<>();
	static List<int[]> virus = new ArrayList<>();
	static Wall[] walls = new Wall[3];
	static Queue<int[]> q = new ArrayDeque<>();
	static boolean[][] vis;
	static int[] dx = {1, 0, -1, 0};
	static int[] dy	= {0, 1, 0, -1};
	static int ans;
	
	public static void main(String[] args) throws Exception {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    		
    	StringTokenizer st = new StringTokenizer(br.readLine());
    	
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	
    	board = new int[N][M];
    	cpy = new int[N][M];
    	vis = new boolean[N][M];
    	
    	for(int i = 0; i < N; i++) {
    		st = new StringTokenizer(br.readLine());
    		
    		for(int j = 0; j < M; j++) {
    			board[i][j] = Integer.parseInt(st.nextToken());
    			if(board[i][j] == 0)
    				blank.add(new int[] {i, j});
    			if(board[i][j] == 2)
    				virus.add(new int[] {i, j});
    		}
    	}
    	
    	for(int i = 0; i < N; i++) {
    		for(int j = 0; j < M; j++)
    			cpy[i][j] = board[i][j];
    	}
    
    	comb(0, 0);
    	
    	System.out.println(ans);
    }
	

	static class Wall {
		int x;
		int y;
		
		Wall(int x, int y){
			this.x = x;
			this.y = y;
		}
	}
	
	static void comb(int srcIdx, int tgtIdx) {
		if(tgtIdx == 3) {
			for(int i = 0; i < 3; i++) {
				cpy[walls[i].x][walls[i].y] = 1;
			}
			
			bfs();
			ans = Math.max(ans, check());
			
			for(int i = 0; i < N; i++) {
	    		for(int j = 0; j < M; j++)
	    			cpy[i][j] = board[i][j];
	    	}
			return;
		}
		
		if(srcIdx == blank.size()) return;
		
		walls[tgtIdx] = new Wall(blank.get(srcIdx)[0], blank.get(srcIdx)[1]);
		comb(srcIdx + 1, tgtIdx + 1);
		comb(srcIdx + 1, tgtIdx);
	}
	
	static void bfs() {
		for(int i = 0; i < N; i++)
			Arrays.fill(vis[i], false);
		
		for(int[] cur : virus) {
			q.offer(cur);
			vis[cur[0]][cur[1]] = true;
		}
		
		while(!q.isEmpty()) {
			int[] cur = q.poll();
			int x = cur[0];
			int y = cur[1];
			
			for(int dir = 0; dir < 4; dir++) {
				int nx = x + dx[dir];
				int ny = y + dy[dir];
				
				if(nx < 0 || nx >= N || ny < 0 || ny >= M) continue;
				if(cpy[nx][ny] != 0 || vis[nx][ny]) continue;
				
				vis[nx][ny] = true;
				cpy[nx][ny] = 2;
				q.offer(new int[] {nx, ny});
			}
		}
	}
	
	static int check() {
		int cnt = 0;
		
		for(int i = 0; i < N; i++) {
			for(int j = 0; j < M; j++) {
				if(cpy[i][j] == 0)
					cnt++;
			}
		}
		
		return cnt;
	}
	
}
